### PR TITLE
[SPARK-19849][SQL] Support ArrayType in to_json to produce JSON array

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to support an array of struct type in `to_json` as below:

```scala
import org.apache.spark.sql.functions._

val df = Seq(Tuple1(Tuple1(1) :: Nil)).toDF("a")
df.select(to_json($"a").as("json")).show()
```

```
+----------+
|      json|
+----------+
|[{"_1":1}]|
+----------+
```

Currently, it throws an exception as below (a newline manually inserted for readability):

```
org.apache.spark.sql.AnalysisException: cannot resolve 'structtojson(`array`)' due to data type 
mismatch: structtojson requires that the expression is a struct expression.;;
```

This allows the roundtrip with `from_json` as below:

```scala
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

val schema = ArrayType(StructType(StructField("a", IntegerType) :: Nil))
val df = Seq("""[{"a":1}, {"a":2}]""").toDF("json").select(from_json($"json", schema).as("array"))
df.show()

// Read back.
df.select(to_json($"array").as("json")).show()
```

```
+----------+
|     array|
+----------+
|[[1], [2]]|
+----------+

+-----------------+
|             json|
+-----------------+
|[{"a":1},{"a":2}]|
+-----------------+
```

Also, this PR proposes to rename from `StructToJson` to `StructsToJson ` and `JsonToStruct` to `JsonToStructs`.

## How was this patch tested?

Unit tests in `JsonFunctionsSuite` and `JsonExpressionsSuite` for Scala, doctest for Python and test in `test_sparkSQL.R` for R.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @brkyvz and @marmbrus, could you please take a look and see if it makes sense?

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out [hunk 1]
```diff
@@ -32,32 +34,40 @@ Usage: to_json(expr[, options]) - Returns a json string with a given struct valu
...
+struct<structorarraytojson(array(named_struct(a, 1, b, 2))):string>
 -- !query 4 output
-org.apache.spark.sql.AnalysisException
-Must use a map() function for options;; line 1 pos 7
+[{"a":1,"b":2}]
 
 
 -- !query 5
-select to_json()
+select to_json(named_struct('a', 1, 'b', 2), named_struct('mode', 'PERMISSIVE'))
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: This happened to look a bit weird but I had to add this in the middle of the sql file - https://github.com/apache/spark/pull/17192/files#diff-3b8a538abd658a260aa32c4aa593bed7R6 to represent this is not the case of the error.

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out [hunk 2]
```diff
@@ -32,32 +34,40 @@ Usage: to_json(expr[, options]) - Returns a json string with a given struct valu
...
 -- !query 3 schema
-struct<structtojson(named_struct(time, to_timestamp('2015-08-26', 'yyyy-MM-dd'))):string>
+struct<structorarraytojson(named_struct(time, to_timestamp('2015-08-26', 'yyyy-MM-dd'))):string>
 -- !query 3 output
 {"time":"26/08/2015"}
 
 
 -- !query 4
-select to_json(named_struct('a', 1, 'b', 2), named_struct('mode', 'PERMISSIVE'))
+select to_json(array(named_struct('a', 1, 'b', 2)))
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: The changed lines happened to look a bit a lot but I had to add this in the middle of the sql file - https://github.com/apache/spark/pull/17192/files#diff-3b8a538abd658a260aa32c4aa593bed7R6 to represent this is not the case of the error.

Actual change mostly here `select to_json(array(named_struct('a', 1, 'b', 2)))`.

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 3]
```diff
@@ -1339,6 +1339,11 @@ test_that("column functions", {
   expect_equal(collect(select(df, bround(df$x, 0)))[[1]][2], 4)
 
   # Test to_json(), from_json()
+  arr <- list(listToStruct(list("name" = "bob")))
+  df <- as.DataFrame(list(listToStruct(list("people" = arr))))
+  j <- collect(select(df, alias(to_json(df$people), "json")))
+  expect_equal(j[order(j$json), ][1], "[{\"name\":\"bob\"}]")
+
```
Comment (HyukjinKwon) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: @felixcheung, it is a little bit of R codes here but could you check the test and documentation?

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out (hunk 2 above)
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: Cc @maropu, thia adds an test in the file you latelly wrote. Could you check if this follows your indention?

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 4]
```diff
@@ -220,4 +242,5 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
     assert(errMsg2.getMessage.startsWith(
       "A type of keys and values in map() must be string, but got"))
   }
+
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: I found a newline at the end of tests prevent a conflict in some cases (_if I am not mistaken_). I am happy to revert this change back if anyone is sure that it is uesless or feel like it is an unrelated change.

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out (hunk 2 above)
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: yea, it looks okay!

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala (hunk 4 above)
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: nit: delete the blank?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 5]
```diff
@@ -422,7 +422,7 @@ object FunctionRegistry {
     expression[BitwiseXor]("^"),
 
     // json
-    expression[StructToJson]("to_json"),
+    expression[StructOrArrayToJson]("to_json"),
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: It seems `StructOrArrayToJson` is a little ambiguous cuz this `Array` means an array of structs, right? `StructsToJson` is better?

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: I think that one is not particually better. `StructsToJson` then does not refer it can be an array or a single struct if the reason is only ambiguosity. Either way is fine to me. If any committer picks up one, let me follow.

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out (hunk 2 above)
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: Thank you for your confirmation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 5 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: okay, thanks!

File: R/pkg/inst/tests/testthat/test_sparkSQL.R (hunk 3 above)
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: this might be a bit hard for regular users to create the DataFrame this way - is there a different way that still produces column that works with `to_json`?

that should probably be in the example in functions.R?

Comment (HyukjinKwon) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Sure, let me add it if I find a better way (it seems `listToStruct` is even an interval API.. ). Is it okay to use `sql("SELECT array(named_struct('name','bob')) as people")` instead if it works fine?

Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: possibly if it does

Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: yes `listToStruct` is internal and it's mucking with types (though it's legitimate to do in R)

Comment (HyukjinKwon) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Let me propose this with it for now and try to find another way until it is merged.

**Comment by HyukjinKwon** [comment]
@brkyvz and @marmbrus, I think it is okay for SQL test and R change and it is ready for a look. Could you take a look please?

File: R/pkg/R/functions.R [hunk 6]
```diff
@@ -1809,8 +1809,13 @@ setMethod("to_date",
...
 #' @examples
 #' \dontrun{
-#' to_json(df$t, dateFormat = 'dd/MM/yyyy')
-#' select(df, to_json(df$t))
+#' # Converts a struct into a JSON object
+#' df <- sql("SELECT named_struct('date', cast('2000-01-01' as date)) as d")
+#' select(df, to_json(df$d, dateFormat = 'dd/MM/yyyy'))
+#'
+#' # Converts an array of structs into a JSON array
+#' df <- sql("SELECT array(named_struct('date', cast('2000-01-01' as date))) as d")
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: nit: maybe add multiple struct in the array instead of just one?

File: python/pyspark/sql/functions.py [hunk 7]
```diff
@@ -1802,10 +1802,10 @@ def from_json(col, schema, options={}):
 @since(2.1)
 def to_json(col, options={}):
     """
-    Converts a column containing a [[StructType]] into a JSON string. Throws an exception,
-    in the case of an unsupported type.
+    Converts a column containing a [[StructType]] or [[ArrayType]] into a JSON string.
+    Throws an exception, in the case of an unsupported type.
```
Comment (felixcheung) [inline python/pyspark/sql/functions.py]: nit: for python and scala here the doc wording can be confusing - it doesn't actually throw an exception right there, by itself with `to_json()`, but it might when it's trying to resolve the column, like with `select(to_json...)`

File: python/pyspark/sql/functions.py [hunk 8]
```diff
@@ -1814,6 +1814,10 @@ def to_json(col, options={}):
     >>> df = spark.createDataFrame(data, ("key", "value"))
     >>> df.select(to_json(df.value).alias("json")).collect()
     [Row(json=u'{"age":2,"name":"Alice"}')]
+    >>> data = [(1, [Row(name='Alice', age=2)])]
```
Comment (felixcheung) [inline python/pyspark/sql/functions.py]: ditto - I'd suggest adding multiple struct in the array

File: python/pyspark/sql/functions.py (hunk 7 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Hm, in that sense, there are many instances, for example, `Returns the double value`, `Returns the positive value` or ``Returns `null` ...``. They return a column. Another example is, `Extracts json object ...`. It does not extract it right there but make a column.

Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: If you don't strongly feel about this, let me leave as is (and sweep it in another PR). In most cases, I believe users will use directly with `select`.

Comment (felixcheung) [inline python/pyspark/sql/functions.py]: that's ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 9]
```diff
@@ -589,52 +591,69 @@ case class StructToJson(
...
+  }
+
+  // This converts rows to the JSON output according to the given schema.
+  @transient
+  lazy val converter: Any => String = {
+    def getAndReset(): String = {
+      gen.flush()
+      val json = writer.toString
+      writer.reset()
+      json
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: nit: why don't you perform `UTF8String.fromString(` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 5 above)
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: since this is an internal name, I'm fine either way, but I prefer `StructsToJson`, because we can't use this method on an array of integers. I feel `StructOrArrayToJson` is a bit more ambiguous

File: python/pyspark/sql/functions.py [hunk 10]
```diff
@@ -1802,10 +1802,10 @@ def from_json(col, schema, options={}):
 @since(2.1)
 def to_json(col, options={}):
     """
-    Converts a column containing a [[StructType]] into a JSON string. Throws an exception,
-    in the case of an unsupported type.
+    Converts a column containing a [[StructType]] or [[ArrayType]] into a JSON string.
```
Comment (brkyvz) [inline python/pyspark/sql/functions.py]: We should mention that the `ArrayType` should be of `StructType`

File: R/pkg/R/functions.R [hunk 11]
```diff
@@ -1795,10 +1795,10 @@ setMethod("to_date",
 
 #' to_json
 #'
-#' Converts a column containing a \code{structType} into a Column of JSON string.
-#' Resolving the Column can fail if an unsupported type is encountered.
+#' Converts a column containing a \code{structType} or an array of \code{structType} into a Column
+#' of JSON string. Resolving the Column can fail if an unsupported type is encountered.
 #'
-#' @param x Column containing the struct
+#' @param x Column containing the struct or an array of the structs
```
Comment (brkyvz) [inline R/pkg/R/functions.R]: nit: `array of structs`

File: python/pyspark/sql/functions.py [hunk 12]
```diff
@@ -1802,10 +1802,10 @@ def from_json(col, schema, options={}):
...
+    Throws an exception, in the case of an unsupported type.
 
-    :param col: name of column containing the struct
+    :param col: name of column containing the struct or an array of the structs
```
Comment (brkyvz) [inline python/pyspark/sql/functions.py]: `array of structs`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala [hunk 13]
```diff
@@ -37,6 +37,11 @@ private[sql] class JacksonGenerator(
 
   // `ValueWriter`s for all fields of the schema
   private val rootFieldWriters: Array[ValueWriter] = schema.map(_.dataType).map(makeWriter).toArray
+  // `ValueWriter` for array data storing rows of the schema.
+  private val arrElementWriter: ValueWriter = {
+    (arr: SpecializedGetters, i: Int) =>
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: ultra nit: I would move this above like:

```scala
private val arrElementWriter: ValueWriter = (arr: SpecializedGetters, i: Int) => {
  ...
}
```

**Review by brkyvz — COMMENTED** [review]
Looks pretty good! Left some minor comments. Thanks for working on this. Would you also like to take a look @hvanhovell ?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
(I just rebased to resolve the conflicts)

**Comment by felixcheung** [comment]
I will do another pass today and merge if others do not have more concerns?

File: python/pyspark/sql/functions.py [hunk 14]
```diff
@@ -1774,10 +1774,11 @@ def json_tuple(col, *fields):
 def from_json(col, schema, options={}):
     """
     Parses a column containing a JSON string into a [[StructType]] or [[ArrayType]]
-    with the specified schema. Returns `null`, in the case of an unparseable string.
+    of [[StructType]]s with the specified schema. Returns `null`, in the case of an unparseable
+    string.
```
Comment (felixcheung) [inline python/pyspark/sql/functions.py]: if we are updating this in python, we should update R too?
https://github.com/HyukjinKwon/spark/blob/185ea6003d60feed20c56de61c17bc304663d99a/R/pkg/R/functions.R#L2440

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 15]
```diff
@@ -425,8 +425,8 @@ object FunctionRegistry {
     expression[BitwiseXor]("^"),
 
     // json
-    expression[StructToJson]("to_json"),
-    expression[JsonToStruct]("from_json"),
+    expression[StructsToJson]("to_json"),
+    expression[JsonToStructs]("from_json"),
```
Comment (felixcheung) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: + @maropu @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 16]
```diff
@@ -624,41 +627,58 @@ case class StructToJson(
...
-      child.dataType.asInstanceOf[StructType],
-      writer,
-      new JSONOptions(options, timeZoneId.get))
+  lazy val gen = new JacksonGenerator(
+    rowSchema, writer, new JSONOptions(options, timeZoneId.get))
+
+  @transient
+  lazy val rowSchema = child.dataType match {
+    case st: StructType => st
+    case ArrayType(st: StructType, _) => st
```
Comment (felixcheung) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: could we end up with `ArrayType(StructType, something_not_struct)` in this match?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 17]
```diff
@@ -624,41 +627,58 @@ case class StructToJson(
...
+          getAndReset()
+    }
+  }
 
   override def dataType: DataType = StringType
 
-  override def checkInputDataTypes(): TypeCheckResult = {
-    if (StructType.acceptsType(child.dataType)) {
+  override def checkInputDataTypes(): TypeCheckResult = child.dataType match {
+    case _: StructType | ArrayType(_: StructType, _) =>
```
Comment (felixcheung) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: this seems to become a bit more strict from before?
was: `StructType.acceptsType` - anything can be accepted as struct
now: `match { case _: StructType` - must be struct

am I understand this correctly?

**Review by felixcheung — COMMENTED** [review]
couple of questions, looks good otherwise

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 16 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Ah, it should be fine. This will be caught in https://github.com/apache/spark/pull/17192/files/185ea6003d60feed20c56de61c17bc304663d99a#diff-6626026091295ad8c0dfb66ecbcd04b1R663 ahead.

File: python/pyspark/sql/functions.py (hunk 14 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Sure, let me try.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 17 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: It seems `StructType.acceptsType` and `ArrayType.acceptsType` call `isInstanceOf[StructType]` and `isInstanceOf[ArrayType]`. `isInstanceOf` and pattern matching are interchangeable in most cases up to my knowledge. 

(I just found a reference https://www.safaribooksonline.com/library/view/scala-cookbook/9781449340292/ch03s15.html)

Namely, this case below should be fine. (Up to my knowledge, Scala forbids case-to-case inheritance BTW)

```
scala> case class A()
defined class A

scala> class B extends A
defined class B

scala> new B() match {case _: A => println(1)}
1
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 15 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: (It was @maropu 's initial suggestion and @brkyvz who could decide what to add agreed on this. It should be fine.)

**Comment by HyukjinKwon** [comment]
Thank you @felixcheung for your review and proceeding this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 17 above)
Comment (felixcheung) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: right, thanks for testing this out.

File: R/pkg/R/functions.R [hunk 18]
```diff
@@ -2438,7 +2438,8 @@ setMethod("date_format", signature(y = "Column", x = "character"),
 #' from_json
 #'
 #' Parses a column containing a JSON string into a Column of \code{structType} with the specified
-#' \code{schema}. If the string is unparseable, the Column will contains the value NA.
+#' \code{schema} or array of \code{structType} if \code{asJsonArray} is enabled. If the string
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: for clarity, I'd suggest saying
`if \code{asJsonArray} is set to \code{TRUE}`

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by felixcheung** [comment]
merged to master

**Comment by HyukjinKwon** [comment]
Thank you @felixcheung, @brkyvz and @maropu.
